hw/ip/otbn/dv/rig/rig/gens/bad_load_store.py

```diff
@@ -0,0 +1,239 @@
+        base = []
+
+        # Get known registers that has out of bounds addresses (from both sides)
+        known_regs = model.regs_with_known_vals('gpr')
+        for reg_idx, reg_val in known_regs:
+            if (min_offset + (1 << 32)) > reg_val > (4096 + min_offset):
+                base.append((reg_idx, reg_val))
+        if not base:
+            return None
+
+        # Pick a random register among known registers that is constrained above
+        idx, value = random.choice(base)
+
+        # If it is close enough, pick top of the memory for "barely" out of bound
+        # coverage point for XW instructions
+        if (4096 + min_offset) < value <= 4096 + max_offset:
+            imm_val = 4096 - value
+        else:
+            imm_val = random.randrange(min_offset, max_offset)
+
+        offset_val = self.imm_op_type.op_val_to_enc_val(imm_val, model.pc)
+        assert offset_val is not None
+
+        # It does not matter for BN.XID instructions
+        bn_imm_val = random.randrange(bn_min_offset, bn_max_offset, 32)
+        bn_offset_val = self.bn_imm_op_type.op_val_to_enc_val(bn_imm_val,
+                                                              model.pc)
+        assert bn_offset_val is not None
+
+        # Get the chosen base register index as grs1 operand.
+        op_val_grs1 = idx
+        if op_val_grs1 is None:
+            return None
+
+        # Any known register is okay for grs2 operand since it will be
+        # guaranteed to have an out of bounds address to store the value from.
+        # We definitely ave a known register (x0), so this should never fail
+        idx_grs2, value_grs2 = random.choice(known_regs)
+        op_val_grs2 = idx_grs2
+        assert op_val_grs2 is not None
+
+        op_vals = []
+        for insn in self.insns:
+            if insn.mnemonic == 'lw':
+                addr = imm_val + value
+                op_vals.append([op_val_grd, offset_val, op_val_grs1])
+            elif insn.mnemonic == 'sw':
+                addr = imm_val + value
+                op_vals.append([op_val_grs2, offset_val, op_val_grs1])
+            elif insn.mnemonic == 'bn.lid':
+                addr = bn_imm_val + value
+                op_vals.append([op_val_grd, op_val_grs1, bn_offset_val, 0, 0])
+            elif insn.mnemonic == 'bn.sid':
+                addr = bn_imm_val + value
+                op_vals.append([op_val_grs1, op_val_grs2, bn_offset_val, 0, 0])
+
+        # Wrap he address if its calculated negative, otherwise we will get
+        # error at the asm stage
+        if addr < 0:
+            addr = addr + (1 << 15)
+        # Pick a random instruction from the weighted list
+        insn_idx = random.choices(range(len(self.insns)),
+                                  weights=self.weights)[0]
+
+        # Weight cannot be zero because in that case we are not adding
+        # it to insns list
+        assert self.weights[insn_idx] > 0
```

Comment:
This is true, but not necessarily needed. The code that does this in `straight_line_insn` is because we mess around with weightings while trying to pick things and there's a danger of making all the weights zero, in which case we can pick something that should have just disabled itself, causing an infinite loop in the generator. Here, that sort of thing can't happen.